### PR TITLE
cli: add lockdown/forward command

### PR DESCRIPTION
**Pull request by doronz88**

### CODE REVIEW

File: pymobiledevice3/service_connection.py [hunk 1]
```diff
@@ -35,37 +35,41 @@ class ServiceConnection(object):
...
-    def connect(self, udid=None):
+    def connect(self):
         mux = usbmux.USBMux()
         mux.process(1.0)
         dev = None
 
         while not dev and mux.devices:
             mux.process(1.0)
-            if udid:
+            if self.udid:
```
Comment (matan1008) [inline pymobiledevice3/service_connection.py]: not None

File: pymobiledevice3/service_connection.py [hunk 2]
```diff
@@ -35,37 +35,41 @@ class ServiceConnection(object):
     def __init__(self, port, udid=None, logger=None):
         self.logger = logger or logging.getLogger(__name__)
         self.port = port
-        self.connect(udid)
+        self.udid = udid
+        self.connect()
 
-    def connect(self, udid=None):
+    def connect(self):
```
Comment (matan1008) [inline pymobiledevice3/service_connection.py]: Wouldn't it be prettier and SOLID to make this function a static method and to pass the socket as an argument to ServiceConnection?

File: pymobiledevice3/lockdown.py [hunk 3]
```diff
@@ -118,6 +120,100 @@ def __init__(self, udid=None, client_name=DEFAULT_CLIENT_NAME):
...
+        # dictionaries containing the required maps to transfer data between each local
+        # socket to its remote socket and vice versa
+        remote_to_local = {}
+        local_to_remote = {}
+
+        # create local tcp server socket
+        server_socket = socket.socket()
+        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        server_socket.bind(('0.0.0.0', src_port))
+        server_socket.listen(200)
```
Comment (matan1008) [inline pymobiledevice3/lockdown.py]: Maybe move to a class or module attribute?

File: pymobiledevice3/lockdown.py [hunk 4]
```diff
@@ -118,6 +120,100 @@ def __init__(self, udid=None, client_name=DEFAULT_CLIENT_NAME):
         self.paired = True
         return
 
+    def forward(self, src_port: int, dst_port: int):
+        """
+        forward each connection from given local machine port to remote device port
+        """
+        # dictionaries containing the required maps to transfer data between each local
+        # socket to its remote socket and vice versa
+        remote_to_local = {}
```
Comment (matan1008) [inline pymobiledevice3/lockdown.py]: Perhaps use the same dictionary for both?

File: pymobiledevice3/lockdown.py [hunk 5]
```diff
@@ -118,6 +120,100 @@ def __init__(self, udid=None, client_name=DEFAULT_CLIENT_NAME):
...
+        local_connection = None
+        remote_connection = None
+
+        while inputs:
+            # will only perform the socket select on the inputs. the outputs will handled
+            # as synchronous blocking
+            readable, writable, exceptional = select.select(inputs, [], inputs)
+
+            for s in readable:
+                if s is server_socket:
```
Comment (matan1008) [inline pymobiledevice3/lockdown.py]: Maybe split the following parts to functions?

**Review by matan1008 — CHANGES_REQUESTED** [review]

File: pymobiledevice3/service_connection.py (hunk 2 above)
Comment (doronz88) [reply pymobiledevice3/service_connection.py]: ok

File: pymobiledevice3/lockdown.py (hunk 3 above)
Comment (doronz88) [reply pymobiledevice3/lockdown.py]: OK

File: pymobiledevice3/lockdown.py (hunk 4 above)
Comment (doronz88) [reply pymobiledevice3/lockdown.py]: ok

File: pymobiledevice3/lockdown.py (hunk 5 above)
Comment (doronz88) [reply pymobiledevice3/lockdown.py]: ok

File: pymobiledevice3/lockdown.py [hunk 6]
```diff
@@ -2,12 +2,14 @@
 import plistlib
 import platform
 import logging
+import select
```
Comment (matan1008) [inline pymobiledevice3/lockdown.py]: Remove

**Review by matan1008 — APPROVED** [review]

**Pull request merged**